src/DataProtection/DataProtection/src/Repositories/EphemeralXmlRepository.cs

```diff
@@ -54,4 +54,59 @@ public void StoreElement(XElement element, string friendlyName)
             _storedElements.Add(cloned);
         }
     }
+
+    /// <inheritdoc/>
+    public bool RemoveElements(Action<IReadOnlyCollection<IDeletableElement>> chooseElements)
+    {
+        ArgumentNullThrowHelper.ThrowIfNull(chooseElements);
+
+        var deletableElements = new List<DeletableElement>();
+
+        foreach (var storedElement in _storedElements)
+        {
+            // Make a deep copy so caller doesn't inadvertently modify it.
+            deletableElements.Add(new DeletableElement(storedElement, new XElement(storedElement)));
+        }
+
+        chooseElements(deletableElements);
+
+        var elementsToDelete = deletableElements
+            .Where(e => e.DeletionOrder.HasValue)
+            .OrderBy(e => e.DeletionOrder.GetValueOrDefault());
+
+        lock (_storedElements)
+        {
+            foreach (var deletableElement in elementsToDelete)
+            {
+                var storedElement = deletableElement.StoredElement;
+                var index = _storedElements.FindIndex(e => ReferenceEquals(e, storedElement));
+                if (index >= 0) // Might not find it if the collection has changed since we started.
+                {
+                    // It would be more efficient to remove the larger indices first, but deletion order
+                    // is important for correctness.
+                    _storedElements.RemoveAt(index);
+                }
```

Comment:
Nit:

```suggestion
                // It would be more efficient to remove the larger indices first, but deletion order
                // is important for correctness.
                _storedElements.Remove(deletableElement.StoredElement);
```

This should still be fine even if we release and reacquire the lock allowing the element to be removed. `Remove` will just return false.

Reply:
Remove uses `object.Equals` and I think we need `object.ReferenceEquals`.